core/nightshade/src/nightshade.rs

```diff
@@ -0,0 +1,781 @@
+
+    #[test]
+    fn test_nightshade_two_authorities() {
+        nightshade_all_sync(2, 5);
+    }
+
+    #[test]
+    fn test_nightshade_three_authorities() {
+        nightshade_all_sync(3, 5);
+    }
+
+    #[test]
+    fn test_nightshade_ten_authorities() {
+        nightshade_all_sync(10, 5);
+    }
+
+    fn bare_state(primary_confidence: i64, endorses: AuthorityId, secondary_confidence: i64) -> BareState {
+        BareState {
+            primary_confidence,
+            endorses: header(endorses),
+            secondary_confidence,
+        }
+    }
+
+    fn state(primary_confidence: i64, endorses: AuthorityId, secondary_confidence: i64) -> State {
+        let mut state = State::empty();
+        state.bare_state = bare_state(primary_confidence, endorses, secondary_confidence);
+        state
+    }
+
+    #[test]
+    fn test_incompatible() {
+        assert_eq!(incompatible_states(&state(4, 1, 2), &state(3, 1, 3)), true);
+        assert_eq!(incompatible_states(&state(4, 1, 3), &state(3, 1, 3)), false);
+        assert_eq!(incompatible_states(&state(4, 2, 2), &state(3, 1, 3)), true);
+        assert_eq!(incompatible_states(&state(4, 2, 2), &state(3, 1, 2)), true);
+    }
+
+    #[test]
+    fn test_order() {
+        // Antisymmetry
+        assert_eq!(bare_state(3, 3, 1) > bare_state(2, 3, 2), true);
+        assert_eq!(bare_state(2, 3, 2) > bare_state(3, 3, 1), false);
+        // No reflexive
+        assert_eq!(bare_state(3, 3, 1) > bare_state(3, 3, 1), false);
+        // Lexicographically correct
+        assert_eq!(bare_state(3, 4, 1) > bare_state(3, 3, 2), true);
+        assert_eq!(bare_state(3, 3, 3) > bare_state(3, 3, 2), true);
+    }
+
+    #[test]
+    fn test_nightshade_basics() {
+        let mut ns = create_nightshades(2);
+
+        let state0 = ns[0].state();
+        assert_eq!(state0.endorses().author, 0);
+        let state1 = ns[1].state();
+        ns[0].update_state(1, state1.clone());
+        let state0 = ns[0].state();
+        assert_eq!(state0.endorses().author, 1);
+    }
+
+    #[test]
+    fn test_nightshade_basics_confidence() {
+        let num_authorities = 4;
+        let mut ns = create_nightshades(num_authorities);
+
+        for i in 0..2 {
+            let state2 = ns[2].state();
+            ns[i].update_state(2, state2);
+            let state_i = ns[i].state();
+            assert_eq!(state_i.endorses().author, 2);
+
+            ns[2].update_state(i, state_i);
+            let state2 = ns[2].state();
+
+            // After update from authority 2 expected confidence is 0 since only authorities 1 and 2
+            // endorse outcome 1. After update from authority 3, there are 3 authorities endorsing 1
+            // with triplet (0, 1, 0) so confidence must be 1.
+            assert_eq!(state2.endorses().author, 2);
+            assert_eq!(state2.bare_state.primary_confidence, i as i64);
+        }
+    }
+
+    // TODO: Tests don't work because of signature verification. Fix them.
```

Comment:
let's leave a tracking bug for this - TODO(issue number).
It helps to figure out who left this / context and triage later.

Reply:
I suggest we add a Nightshade page to nomicon where we track things that we have and things that are missing using bullet lists, similar to network, see: https://nearprotocol.github.io/nomicon/Network/Network.html and https://github.com/nearprotocol/nomicon .

Eventually that's where the spec of Nightshade will be, so I suggest starting with missing/existing features list.

Reply:
I'll use Nomicon to track consensus related issues. This particular TODO is deprecated since all commented tests won't work until tests are well designed. I will work on tests ASAP since this code needs to be bulletproof, regarding security & performance.